File: leetcode/golang/232_implement_queue_using_stacks.go
type MyQueue struct {
    A []int
    B []int
}


func Constructor() MyQueue {
    return MyQueue{}
}


func (this *MyQueue) Push(x int)  {
    this.A = append(this.A, x)
}


func (this *MyQueue) Pop() int {
    peek := this.Peek()
    if peek != -1 {
        this.B = this.B[:len(this.B) - 1]
    }
    return peek
}


func (this *MyQueue) Peek() int {
    if (len(this.B) != 0) {
        return this.B[len(this.B) - 1]
    }
    if (len(this.A) == 0) {
        return -1
    }
    for len(this.A) != 0 {
        this.B = append(this.B, this.A[len(this.A) - 1])
        this.A = this.A[:len(this.A) - 1]
    }
    return this.B[len(this.B) - 1]
}


func (this *MyQueue) Empty() bool {
    return len(this.A) == 0 && len(this.B) == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */

File: leetcode/c++/232_implement_queue_using_stacks.cpp
// https://leetcode.cn/problems/implement-queue-using-stacks/description

#include <bits/stdc++.h>
using namespace std;

class MyQueue {
private:
    stack<int>A, B;
public:
    MyQueue() {

    }
    
    void push(int x) {
       A.push(x);
    }
    
    int pop() {
        int peek = this->peek();
        if ( peek!= -1 ) 
            B.pop();        
        return peek;
    }
    
    int peek() {
        if (!B.empty())
            return B.top();
        if (A.empty())
            return -1;
        while (!A.empty()) {
            B.push(A.top());
            A.pop();
        }
        return B.top();
    }
    
    bool empty() {
        return A.empty() && B.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

// 时间复杂度为O(1), 空间复杂度为O(n)